nebula2/mclient/__init__.py

```diff
@@ -0,0 +1,392 @@
+            req = ListHostsReq()
+            # req.role = HostRole.STORAGE
+            resp = self._connection.listHosts(req)
+            if resp.code != ErrorCode.SUCCEEDED:
+                self.update_leader()
+                raise RuntimeError("List spaces failed, error code: {}"
+                                   .format(ErrorCode._VALUES_TO_NAMES(resp.code)))
+            return resp.hosts
+
+    def get_parts_alloc(self, space_id):
+        with self._lock:
+            if self._connection is None:
+                raise RuntimeError('The connection is no open')
+            req = GetPartsAllocReq()
+            req.space_id = space_id
+            resp = self._connection.getPartsAlloc(req)
+            if resp.code != ErrorCode.SUCCEEDED:
+                self.update_leader()
+                raise RuntimeError("List parts from space id:{} failed, error code: {}"
+                                   .format(space_id, ErrorCode._VALUES_TO_NAMES(resp.code)))
+            return resp.parts
+
+    def close(self):
+        try:
+            if self._connection is not None:
+                self._connection._iprot.trans.close()
+        except Exception:
+            raise
+
+    def update_leader(self, resp):
+        if resp.code == ErrorCode.E_LEADER_CHANGED:
+            try:
+                if resp.leader is not None:
+                    with self._lock:
+                        self._leader = (resp.leader.host, resp.leader.port)
+                        self.open()
+            except Exception as e:
+                logging.error(e)
+
+    def __del__(self):
+        self.close()
+
+
+class MetaCache(object):
+    class SpaceCache:
+        def __init__(self):
+            self.space_id = 0
+            self.space_name = ''
+            self.tag_items = {}
+            self.edge_items = {}
+            self.parts_alloc = {}
+
+        def __repr__(self):
+            return 'space_id: {}, space_name: {}, tag_items: {}, edge_items: {}, parts_alloc: {}'\
+                .format(self.space_id,
+                        self.space_name,
+                        self.tag_items,
+                        self.edge_items,
+                        self.parts_alloc)
+
+    def __init__(self, meta_addrs, timeout=2000, load_period=10, decode_type='utf-8'):
+        self._decode_type = decode_type
+        self._load_period = load_period
+        self._lock = RLock()
+        self._meta_client = MetaClient(meta_addrs, timeout)
+        self._meta_client.open()
+        self._space_caches = {}
+        self._space_id_names = {}
+        self._storage_addrs = []
+        self._storage_leader = {}
+        self._close = False
+
+        # load meta data
+        self._load_all()
+
+    def __del__(self):
+        self._close = True
+
+    def _period_update_meta_data(self):
+        if self._close:
+            return
+        self._load_all()
+        timer = threading.Timer(self._load_period, self._period_update_meta_data)
```

Comment:
It will produce a Timer object every time when this timer task is executed. Is it suitable?

Reply:
There is no place to use this, it is to delete,